server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,439 @@
+        UPDATE hdb_catalog.hdb_scheduled_events
+        SET delivered = 't', locked = 'f'
+        WHERE id = $1
+      |] (Identity $ sefId se) True
+
+processDead :: (MonadIO m) => Q.PGPool -> ScheduledEventFull -> m (Either QErr ())
+processDead pgpool se =
+  liftIO $
+  runExceptT $ Q.runTx pgpool (Q.RepeatableRead, Just Q.ReadWrite) markDead
+  where
+    markDead =
+      Q.unitQE
+        defaultTxErrorHandler
+        [Q.sql|
+          UPDATE hdb_catalog.hdb_scheduled_events
+          SET dead = 't', locked = 'f'
+          WHERE id = $1
+        |] (Identity $ sefId se) False
+
+setRetry :: ScheduledEventFull -> UTCTime -> Q.TxE QErr ()
+setRetry se time =
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+          UPDATE hdb_catalog.hdb_scheduled_events
+          SET next_retry_at = $1, locked = 'f'
+          WHERE id = $2
+          |] (time, sefId se) True
+
+mkInvocation
+  :: ScheduledEventFull -> Int -> [HeaderConf] -> TBS.TByteString -> [HeaderConf]
+  -> Invocation
+mkInvocation se status reqHeaders respBody respHeaders
+  = let resp = if isClientError status
+          then mkClientErr respBody
+          else mkResp status respBody respHeaders
+    in
+      Invocation
+      (sefId se)
+      status
+      (mkWebhookReq (J.toJSON se) reqHeaders invocationVersion)
+      resp
+
+insertInvocation :: Invocation -> Q.TxE QErr ()
+insertInvocation invo = do
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+          INSERT INTO hdb_catalog.hdb_scheduled_event_invocation_logs
+          (event_id, status, request, response)
+          VALUES ($1, $2, $3, $4)
+          |] ( iEventId invo
+             , toInt64 $ iStatus invo
+             , Q.AltJ $ J.toJSON $ iRequest invo
+             , Q.AltJ $ J.toJSON $ iResponse invo) True
+  Q.unitQE defaultTxErrorHandler [Q.sql|
+          UPDATE hdb_catalog.hdb_scheduled_events
+          SET tries = tries + 1
+          WHERE id = $1
+          |] (Identity $ iEventId invo) True
+
+getScheduledEvents :: Q.TxE QErr [ScheduledEventPartial]
+getScheduledEvents = do
+  partialSchedules <- map uncurryEvent <$> Q.listQE defaultTxErrorHandler [Q.sql|
+      UPDATE hdb_catalog.hdb_scheduled_events
+      SET locked = 't'
+      WHERE id IN ( SELECT t.id
+                    FROM hdb_catalog.hdb_scheduled_events t
+                    WHERE ( t.locked = 'f'
+                            and t.cancelled = 'f'
+                            and t.delivered = 'f'
+                            and t.error = 'f'
+                            and (
+                             (t.next_retry_at is NULL and t.scheduled_time <= now()) or
+                             (t.next_retry_at is not NULL and t.next_retry_at <= now())
+                            )
+                            and t.dead = 'f'
+                          )
+                    FOR UPDATE SKIP LOCKED
+                    )
+      RETURNING id, name, scheduled_time, additional_payload, tries
+      |] () True
+  pure $ partialSchedules
```

Comment:
This bind followed by `pure` is redundant, since `do { x <- e; pure x }` is always the same as just writing `e` (by the monad laws).